### PR TITLE
automate vm_image_url and virtual_uc_ctlpane_interface

### PR DESCRIPTION
**Pull request by Shrutikabansal**

### CODE REVIEW

**Comment by masco** [comment]
Can one of the admins verify this patch?

**Comment by Shrutikabansal** [comment]
@asyedham  could you please review this one.

File: virtual_undercloud.yml [hunk 1]
```diff
@@ -19,14 +19,31 @@
...
+      retries: 3
+      ignore_errors: yes
+      when: osp_release > 13
 
     - name: create undercloud vm
       shell: |
          source {{ infrared_dir }}/.venv/bin/activate
-         infrared virsh --host-address 127.0.0.1 --host-key /root/.ssh/id_rsa --image-url '{{ vm_image_url }}' --topology-network virtual_uc_network --topology-nodes virtual_uc_topology:1 --host-memory-overcommit True -vvv 2>&1 | tee log_virsh1
+         infrared virsh --host-address 127.0.0.1 --host-key /root/.ssh/id_rsa --image-url '{{ vm_image_url.rhel7 }}' --topology-network virtual_uc_network --topology-nodes virtual_uc_topology:1 --host-memory-overcommit True -vvv 2>&1 | tee log_virsh1
+      when: osp_release == 13
```
Comment (asyedham) [inline virtual_undercloud.yml]: condition should be osp_release <= 13

File: virtual_undercloud.yml [hunk 2]
```diff
@@ -19,14 +19,31 @@
...
+
+    - name: create undercloud vm
+      shell: |
+         source {{ infrared_dir }}/.venv/bin/activate
+         infrared virsh --host-address 127.0.0.1 --host-key /root/.ssh/id_rsa --image-url '{{ vm_image_url.rhel8 }}' --topology-network virtual_uc_network --topology-nodes virtual_uc_topology:1 --host-memory-overcommit True -vvv 2>&1 | tee log_virsh1
```
Comment (asyedham) [inline virtual_undercloud.yml]: add a task  to create VM for OSP13 or below

File: tasks/load_instackenv.yml [hunk 3]
```diff
@@ -37,6 +37,10 @@
       set_fact:
         undercloud_hostname: "{{ undercloud_hostname|default(instackenv_content.nodes[0].pm_addr | replace('mgmt-','') | replace('-drac', '')) }}"
 
+    - name: for machine type
```
Comment (asyedham) [inline tasks/load_instackenv.yml]: why is this needed?

File: tasks/load_instackenv.yml [hunk 4]
```diff
@@ -50,6 +54,14 @@
       set_fact:
         undercloud_hostname: "{{ undercloud_host }}"
 
+    - name: undercloud_interface
+      set_fact:
+        undercloud_interface: "{{ undercloud_interface|default(instackenv_content.nodes[0].pm_addr | replace('mgmt-','') | replace('-drac', '')) }}"
+
+    - name: machine type when virtual indercloud true
```
Comment (asyedham) [inline tasks/load_instackenv.yml]: same here

**Review by asyedham — CHANGES_REQUESTED** [review]

File: tasks/load_instackenv.yml (hunk 3 above)
Comment (Shrutikabansal) [reply tasks/load_instackenv.yml]: because used machine type is fine if virtual_uc is false

File: virtual_undercloud.yml (hunk 2 above)
Comment (Shrutikabansal) [reply virtual_undercloud.yml]: i wrote it ==13 which is for below, i'll correct it

File: tasks/load_instackenv.yml [hunk 5]
```diff
@@ -50,6 +54,14 @@
       set_fact:
         undercloud_hostname: "{{ undercloud_host }}"
 
+    - name: undercloud_interface
+      set_fact:
+        undercloud_interface: "{{ undercloud_interface|default(instackenv_content.nodes[0].pm_addr | replace('mgmt-','') | replace('-drac', '')) }}"
```
Comment (masco) [inline tasks/load_instackenv.yml]: I don't think it will return an interface

File: virtual_undercloud.yml [hunk 6]
```diff
@@ -19,14 +19,31 @@
     - name: clean previous vm
       shell: |
          source {{ infrared_dir }}/.venv/bin/activate
-         infrared virsh --host-address 127.0.0.1 --host-key /root/.ssh/id_rsa --image-url '{{ vm_image_url }}' --topology-network virtual_uc_network --topology-nodes virtual_uc_topology:1 --host-memory-overcommit True --cleanup True
+         infrared virsh --host-address 127.0.0.1 --host-key /root/.ssh/id_rsa --image-url '{{ vm_image_url.rhel7 }}' --topology-network virtual_uc_network --topology-nodes virtual_uc_topology:1 --host-memory-overcommit True --cleanup True
```
Comment (masco) [inline virtual_undercloud.yml]: add a task for setting the image url and pass that to the previous tasks, no need to add a clean and create vm for each rhel image. it will be more clean and easy to maintain

File: virtual_interface.yml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+          set_fact:
+            ifaces_v: "{{ item.value.interfaces }}"
+          with_dict: "{{ alias.machine_types.dell }}"
+          when: item.key == machine_type and machine_type in alias.machine_types.dell
+        when: lab_name in ['scale', 'alias']
+
+
+      - name: set interface
+        set_fact:
+          virtual_uc_ctlplane_interface: "{{ ifaces_v[1] }}"
```
Comment (masco) [inline virtual_interface.yml:28]: where u populating the ifaces_v for non alias and scale lab? for example ci lab

File: virtual_interface.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+      - name: print name of interface
```
Comment (masco) [inline virtual_interface.yml]: remove the debug messages

**Review by masco — COMMENTED** [review]
Hey, excellent work as a newly joined member. I have suggested some changes for good code practice. if you have any doubt about that let me know. thanks!

**Comment by Shrutikabansal** [comment]
@masco i did changes for vm automation
still working on interface

**Comment by masco** [comment]
shruti, it is ok. BTW you can divide this PR in to two,  1. VM automation 2.Interface selection. It is good practice to keep one functionality in one PR. It is easy for reviewers and your patch also get merged soon

**Comment by Shrutikabansal** [comment]
Okay , Thank you masco

**Comment by Shrutikabansal** [comment]
I'll be creating two separate pull requests.
